# Twitter Text Preprosessing Overlook

This notebook is to try out the preprocessing on the input files.
The after the tokeniser function finalised it is been put into the Preprocessor Module for use. The notebook then import Preprocessor.py to the notebook and test the functions on in the notebook

Steps:
Preprocessing:
* tokenise http://www.nltk.org/api/nltk.tokenize.html
* lower case
* Remove puntuation, regonise links, AT users, 
* option to regonise/remove stop words
* stemming and lemmatisation http://textminingonline.com/dive-into-nltk-part-iv-stemming-and-lemmatization
* pick out emoticons
* remove duplication characters (done as part of TwitterTokenizer)


Feature extracting:
* word exestence and count
* tf, tfidf
* count for hashtag, at, and link
* n-gram: bigram, trigram and skipgram
* lexicons and 
* word embedding

http://www.nltk.org/api/nltk.sentiment.html  

The tokenise function and extract_features function both have some arguments to control part of the preprocessing so we can change some parameters during text classification to find the best option.

### There is a stopwords corpus in nltk (installed by nltk.download()):

In [1]:
import nltk
from nltk.corpus import stopwords

print stopwords.words('english')[0:30]

# nltk.download(corpus)

[u'i', u'me', u'my', u'myself', u'we', u'our', u'ours', u'ourselves', u'you', u'your', u'yours', u'yourself', u'yourselves', u'he', u'him', u'his', u'himself', u'she', u'her', u'hers', u'herself', u'it', u'its', u'itself', u'they', u'them', u'their', u'theirs', u'themselves', u'what']


### Trying and testing the tokenise function:

In [2]:
import nltk
import re
from nltk.tokenize import TweetTokenizer
from nltk.corpus import stopwords as stopwordloader
from nltk.stem import WordNetLemmatizer
import string

t1 = "@Harry_Styles is so damn cute!!! 1 day to go DIRECTIONERS!!!!!!!! It's coming out at 8"
t2 = "I have one awesome boyfriend who Sat and watched jersey shore with me just to make me happy :)"
t3 = "We're going to be in the rose parade tomorrow with the other obedience club dogs! Lizzie will be wearing roses! :-)"
t4 = "New header of Lana Del Rey *__* her face is getting blocked though, I may change it later.."
t5 = "she call me baby, I call her bae, sorry 4 the wait, Carter IV on the way! august 29th. hellllll yeahhh! (;"
t6 = "RT @coupleBOOBS: Nicki Minaj - Pink Friday, up 15 places, to #16 on the Official UK Top 40 :) http://twitpic.com/62iohc -__-"
t7 = "@kjrmitch you should call in sick....i mean who can listen to HM tomorrow.  Blah, blah...matt Flynn...blah, blah blah...matt Flynn....blah,"
t8 = "Going to the UCLA football game this saturday with my boo . &lt;3  ^.^\n"
t9 = "sooo i might go to laidback luke on friday with nessaaaa ^.^ we going hard &gt;:]\n"

# the tokenise function is written in Preprocessor.py, we import it here
from Preprocessor import tokenise

tokenise(t1, begin=1, end = 3)

[u'is', u'so', u'damn']

In [3]:
# more testing 
print tokenise(t2)
print tokenise(t3)
print tokenise(t4)
print tokenise(t5)
print tokenise(t6)
print tokenise(t7)
print tokenise(t8)
print tokenise(t9)

[u'I', u'have', u'one', u'awesome', u'boyfriend', u'who', u'sat', u'and', u'watched', u'jersey', u'shore', u'with', u'me', u'just', u'to', u'make', u'me', u'happy', u':)']
[u"we're", u'going', u'to', u'be', u'in', u'the', u'rose', u'parade', u'tomorrow', u'with', u'the', u'other', u'obedience', u'club', u'dogs', u'!', u'lizzie', u'will', u'be', u'wearing', u'roses', u'!', u':-)']
[u'new', u'header', u'of', u'lana', u'del', u'rey', u'*__*', u'her', u'face', u'is', u'getting', u'blocked', u'though', u'I', u'may', u'change', u'it', u'later', u'..']
[u'she', u'call', u'me', u'baby', u'I', u'call', u'her', u'bae', u'sorry', u'NUMBER', u'the', u'wait', u'carter', u'IV', u'on', u'the', u'way', u'!', u'august', u'NUMBER', u'helll', u'yeahhh', u'!', u'(;']
[u'RT', u'AT_USER', u'nicki', u'minaj', u'pink', u'friday', u'up', u'NUMBER', u'places', u'to', u'#16', u'on', u'the', u'official', u'UK', u'top', u'NUMBER', u':)', u'LINK', u'-__-']
[u'AT_USER', u'you', u'should', u'call', u'in', u'sick', u'

### Add support for Japanese style Emoticons
the nltk TweetTokenizer which our tokeniser is based on work well in picking up western style emoticons, however it does not cover Japanese style Emoticons (^.^ , .__. , ^^ , etc)  
The nltk TweetTokenizer is well written and structured so I took it and customise it to add regex for Japanese style Emoticons.

In [4]:
import py_compile
py_compile.compile("custom_tokenize.py")

import Preprocessor
from Preprocessor import tokenise

print tokenise("some *__* and ^.^, .__., .___. also .________.   >_< ^^")

[u'some', u'*__*', u'and', u'^.^', u'.__.', u'.___.', u'also', u'.___.', u'>_<', u'^^']


In [5]:
tokenise("[Picture, Translation] Sina Interview with Hyung Jun ^^video thanks to hyungjunthebest@youtube http://t.co/x5TLedQl")

[u'picture',
 u'translation',
 u'sina',
 u'interview',
 u'with',
 u'hyung',
 u'jun',
 u'^^',
 u'video',
 u'thanks',
 u'to',
 u'hyungjunthebest',
 u'AT_USER',
 u'LINK']

In [6]:
EMOTICONS = r"""
    (?:
      [<>]?
      [:;=8]                     # eyes
      [\-o\*\']?                 # optional nose
      [\)\]\(\[dDpP/\:\}\{@\|\\] # mouth
      |
      [\)\]\(\[dDpP/\:\}\{@\|\\] # mouth
      [\-o\*\']?                 # optional nose
      [:;=8]                     # eyes
      [<>]?
      |
      <3                         # heart
      |
      [\*#\O~\-=>][_\.]+[\*#\O~\-=<]      # Japanese style Emoticons eye mouth eye
      |
      [\.][_]+[\.]               # poker face
      |
      \^[_.;]?\^                 # smiling eyes
    )"""

print re.compile(EMOTICONS, re.VERBOSE | re.I | re.UNICODE ).match('^^')
    

In [7]:
print re.match(r"@.*", '@someone')
print re.match(r"#.*", '#sdf')
print re.match(r"http.*", 'https://sdfdsf')
print re.match(r"http.*", 'sdfhttp://')

None


### Strip handles for @user and http

In [8]:
tokenise("@user this is AWESOME!!1Let's get the#party started!http://someurl.com")

[u'AT_USER',
 u'this',
 u'is',
 u'AWESOME',
 u'!',
 u'!',
 u'NUMBER',
 u's',
 u'get',
 u'the',
 u'#party',
 u'started',
 u'!',
 u'LINK']

Tokenise problem:  
we can tokenise smilies like :),   
but we cannot process things like ^.^ and "\-\_\_\-",  
but we tokenise any emoticon as \_\_, which might not be a bad thing

### Let's examine the most common words in the training set B

In [9]:
import csv
from collections import Counter 

tweets={'positive':[], 'negative':[], 'neutral':[]}
dictionary = {'positive':Counter(), 'negative':Counter(), 'neutral':Counter()}
word_dict=Counter()

tfile = "Exercise2_data/twitter-train-cleansed-B.tsv"
with open(tfile) as tsvfile:
    tsvreader = csv.reader(tsvfile, delimiter="\t")
    for line in tsvreader:
        tclass = line[2]
        tweet = line[3]
        tweets[tclass].append(tweet)
        tokens = tokenise(tweet)   
        dictionary[tclass].update(tokens)
        word_dict.update(tokens)

for key, value in dictionary.iteritems():
    print "most common in", key, value.most_common(20)


most common in positive [(u'!', 2582), (u'the', 2373), (u'NUMBER', 1796), (u'to', 1413), (u'AT_USER', 1387), (u'in', 905), (u'on', 874), (u'and', 829), (u'I', 764), (u'a', 764), (u'for', 733), (u'of', 693), (u'you', 629), (u'tomorrow', 589), (u'is', 583), (u'at', 558), (u'...', 523), (u'LINK', 515), (u'be', 507), (u'with', 484)]
most common in neutral [(u'the', 3444), (u'NUMBER', 3038), (u'to', 1764), (u'AT_USER', 1517), (u'in', 1326), (u'on', 1291), (u'LINK', 1238), (u'!', 1219), (u'...', 1026), (u'and', 999), (u'of', 981), (u'for', 890), (u'a', 883), (u'at', 873), (u'with', 695), (u'?', 677), (u'I', 656), (u'is', 653), (u'tomorrow', 609), (u'you', 533)]
most common in negative [(u'the', 918), (u'NUMBER', 602), (u'to', 524), (u'AT_USER', 388), (u'I', 387), (u'a', 360), (u'in', 325), (u'!', 321), (u'on', 313), (u'and', 309), (u'is', 298), (u'of', 277), (u'...', 262), (u'?', 246), (u'for', 216), (u'be', 193), (u'tomorrow', 182), (u'it', 178), (u'you', 174), (u'with', 161)]


### Examing the windows in  some tweets in task A

In [10]:
import linecache
import functools

def inspectTweetsA(filepath, linenumbers, tokeniser):
    tweets = []
    for linenumber in linenumbers:
        line = linecache.getline(filepath, linenumber+1).split('\t')
        begin = int(line[2])
        end = int(line[3])
        label = line[4]
        tokens = tokeniser(line[5], begin=begin, end=end)
        tweets.append((linenumber, begin, end, label, tokens))
    return tweets    

tokenise_func = functools.partial(tokenise , more_instances=0)
twitter_train_A_path = "Exercise2_data/twitter-train-cleansed-A.tsv"

tweets = inspectTweetsA(twitter_train_A_path, range(700,720), tokenise_func)
tweets

[(700, 1, 1, 'positive', [u'congratulations']),
 (701, 13, 13, 'positive', [u'happy']),
 (702, 9, 9, 'positive', [u':D']),
 (703, 19, 19, 'positive', [u':)']),
 (704, 23, 23, 'negative', [u'DEATH']),
 (705, 5, 5, 'negative', [u'worst']),
 (706, 1, 4, 'neutral', [u'is', u'going', u'to', u'change']),
 (707, 20, 20, 'neutral', [u'as']),
 (708, 11, 11, 'negative', [u'for']),
 (709, 6, 7, 'positive', [u'fun-filled', u'day']),
 (710, 0, 2, 'positive', [u'so', u'proud', u'of']),
 (711, 7, 7, 'negative', [u'upset']),
 (712, 3, 3, 'positive', [u'great']),
 (713, 2, 2, 'positive', [u'lucky']),
 (714, 9, 12, 'positive', [u'DELIGHTED', u'what', u'a', u'SATURDAY']),
 (715, 14, 14, 'positive', []),
 (716, 2, 2, 'positive', [u'intelligent']),
 (717, 5, 5, 'positive', [u'mindfulness']),
 (718, 12, 12, 'positive', [u'glorious']),
 (719, 17, 19, 'positive', [u'congratulations', u'to', u'egypt'])]

**Notice** that tweet number 715 does not contain anything at word 14,  
the tweet is "I just remembered 'Niggas In Paris' is a song. DELIGHTED! What a SATURDAY!  #positivetroll"  

There are only 13 words in the tweet ignoring puncuation(as all the other lines) so it must have been an error in the training set.  

I left it being empty because during training, if it does not contain any feature, it would not enter the for loop "for fname in featureset:" in all the classifier we are using, so it would not be counted, which is what we wanted.

For the tokeniser, we add option to add before and after 1 token and it prove to improve performance of the classifier

## K skip bigram

In [11]:
from sklearn.base import BaseEstimator, TransformerMixin
class extract_kskip_bigram(BaseEstimator, TransformerMixin):
    
    def __init__(self, skip=1, hashing=False):
        self.skip = skip
        self.hashing = hashing

    def fit(self, x, y=None):
        return self

    def transform(self, tokenlist):
        featurelist = []
        
        for tokens in tokenlist:
            features = {}  
            tokens = [ word.strip('#') for word in tokens ] #strip hashtag
            
            for s in range(self.skip+1):
            
                ngrams = zip(*[tokens[i:] for i in (0,1+s)])
                for tup in ngrams:
                    key = ' '.join(tup)
                    if self.hashing:
                        features[key] = True
                    else:
                        features[key] = features.get(key, 0) +1
            featurelist.append(features)

        return featurelist

In [12]:
tokenlist = [tokenise(t3),tokenise(t4)]
print tokenlist
# for tokens in tokenlist:
#     print zip(*[tokens[i:] for i in range(2)]).count((u"we're", u'going'))

extract_kskip_bigram(skip=1).transform( tokenlist)

[[u"we're", u'going', u'to', u'be', u'in', u'the', u'rose', u'parade', u'tomorrow', u'with', u'the', u'other', u'obedience', u'club', u'dogs', u'!', u'lizzie', u'will', u'be', u'wearing', u'roses', u'!', u':-)'], [u'new', u'header', u'of', u'lana', u'del', u'rey', u'*__*', u'her', u'face', u'is', u'getting', u'blocked', u'though', u'I', u'may', u'change', u'it', u'later', u'..']]


[{u'! :-)': 1,
  u'! lizzie': 1,
  u'! will': 1,
  u'be in': 1,
  u'be roses': 1,
  u'be the': 1,
  u'be wearing': 1,
  u'club !': 1,
  u'club dogs': 1,
  u'dogs !': 1,
  u'dogs lizzie': 1,
  u'going be': 1,
  u'going to': 1,
  u'in rose': 1,
  u'in the': 1,
  u'lizzie be': 1,
  u'lizzie will': 1,
  u'obedience club': 1,
  u'obedience dogs': 1,
  u'other club': 1,
  u'other obedience': 1,
  u'parade tomorrow': 1,
  u'parade with': 1,
  u'rose parade': 1,
  u'rose tomorrow': 1,
  u'roses !': 1,
  u'roses :-)': 1,
  u'the obedience': 1,
  u'the other': 1,
  u'the parade': 1,
  u'the rose': 1,
  u'to be': 1,
  u'to in': 1,
  u'tomorrow the': 1,
  u'tomorrow with': 1,
  u"we're going": 1,
  u"we're to": 1,
  u'wearing !': 1,
  u'wearing roses': 1,
  u'will be': 1,
  u'will wearing': 1,
  u'with other': 1,
  u'with the': 1},
 {u'*__* face': 1,
  u'*__* her': 1,
  u'I change': 1,
  u'I may': 1,
  u'blocked I': 1,
  u'blocked though': 1,
  u'change it': 1,
  u'change later': 1,
  u'del *__*':

In [13]:
tokens = tokenlist[0]
print tokens
skip = 2
for s in range(0, skip+1):
    print zip(*[tokens[i:] for i in (0,1+s)])

[u"we're", u'going', u'to', u'be', u'in', u'the', u'rose', u'parade', u'tomorrow', u'with', u'the', u'other', u'obedience', u'club', u'dogs', u'!', u'lizzie', u'will', u'be', u'wearing', u'roses', u'!', u':-)']
[(u"we're", u'going'), (u'going', u'to'), (u'to', u'be'), (u'be', u'in'), (u'in', u'the'), (u'the', u'rose'), (u'rose', u'parade'), (u'parade', u'tomorrow'), (u'tomorrow', u'with'), (u'with', u'the'), (u'the', u'other'), (u'other', u'obedience'), (u'obedience', u'club'), (u'club', u'dogs'), (u'dogs', u'!'), (u'!', u'lizzie'), (u'lizzie', u'will'), (u'will', u'be'), (u'be', u'wearing'), (u'wearing', u'roses'), (u'roses', u'!'), (u'!', u':-)')]
[(u"we're", u'to'), (u'going', u'be'), (u'to', u'in'), (u'be', u'the'), (u'in', u'rose'), (u'the', u'parade'), (u'rose', u'tomorrow'), (u'parade', u'with'), (u'tomorrow', u'the'), (u'with', u'other'), (u'the', u'obedience'), (u'other', u'club'), (u'obedience', u'dogs'), (u'club', u'!'), (u'dogs', u'lizzie'), (u'!', u'will'), (u'lizzie', u'b

## Using sentiwordnet from nltk

In [14]:
from nltk.corpus import sentiwordnet as swn
import numpy as np

word = 'late'

print max([x.pos_score() for x in list(swn.senti_synsets(word))])
print max([x.neg_score() for x in list(swn.senti_synsets(word))])
print max([x.obj_score() for x in list(swn.senti_synsets(word))])

print np.mean([x.pos_score() for x in list(swn.senti_synsets(word))])
print np.mean([x.neg_score() for x in list(swn.senti_synsets(word))])
print np.mean([x.obj_score() for x in list(swn.senti_synsets(word))])

0.125
0.25
1.0
0.0178571428571
0.0357142857143
0.946428571429


In [15]:
class lexicon_sentiwordnet(BaseEstimator, TransformerMixin):
    
    def __init__(self, feature='all'):
        self.feature = feature

    def fit(self, x, y=None):
        return self

    # helper method
    def get_lexi_feature(self, features):
        senti_list = lambda word: list(swn.senti_synsets(word)) 
        pos = sum([ max([0]+[ x.pos_score() for x in senti_list(word) ])*features[word] for word in features ])
        neg = sum([ max([0]+[ x.neg_score() for x in senti_list(word) ])*features[word] for word in features ])
        obj = sum([ max([0]+[ x.obj_score() for x in senti_list(word) ])*features[word] for word in features ])
        if self.feature == 'all':
            return {'swn_pos': pos, 'swn_neg': neg, 'swn_obj': obj }
        else:
            return {'swn_score': pos - neg }
            
    def transform(self, featurelist):
        return [ self.get_lexi_feature(features) for features in featurelist ]


In [16]:
featurelist = [{
  u'be': 1,
  u'lizzie': 1,
  u'will': 1,
  u'!': 1,
  u'in': 1,
  u'rose': 1,
  u'roses': 1,
  u'the': 1,
  u'wearing': 1,
  u'club': 1,
  u'going': 1,
  u'parade': 1},{
  u'rose': 1,
  u'obedience': 1,
  u'other': 1,
  u'tomorrow': 1,
        u'happy': 1,
}]

lexicon_sentiwordnet(feature='score').transform(featurelist)

[{'swn_score': 0.625}, {'swn_score': 0.75}]

## Using Liu Hu Lexicon from nltk

In [17]:
from sklearn.base import BaseEstimator, TransformerMixin
from nltk.corpus import opinion_lexicon
from multiprocessing import Pool

class lexicon_liuhu(BaseEstimator, TransformerMixin):

    def __init__(self):
        self.pos_lexicon = set(opinion_lexicon.positive())
        self.neg_lexicon = set(opinion_lexicon.negative())
    
    def fit(self, x, y=None):
        return self
            
    def transform(self, featurelist):
        def get_pos_neg(features):
            pos = sum(1 for word in features if word in self.pos_lexicon)
            neg = sum(1 for word in features if word in self.neg_lexicon)
            return [pos, neg]
        
        return [ get_pos_neg(features) for features in featurelist ]

In [18]:
timeit lexicon_liuhu().transform(featurelist)

100 loops, best of 3: 17.5 ms per loop


## Sentiment140 and NRC Hashtag Lexicon Bigram

In [19]:
import Preprocessor
import PipelineRunner

reload(Preprocessor)
gold_set = PipelineRunner.getTrainingSetB("Exercise2_data/twitter-dev-gold-B.tsv", tokenise)
lexi = Preprocessor.extract_kskip_bigram(skip=3).transform(gold_set['tokens'])
lexi = Preprocessor.lexicon_NRC_bigram().transform(lexi)

In [20]:
len([ (i, k) for i, k in enumerate(lexi) if sum(k)!=0 ] )

1374

In [21]:
gold_set['tokens'][1033]

[u'will',
 u'take',
 u'that',
 u'result',
 u'top',
 u'the',
 u'group',
 u'as',
 u'well',
 u'bet',
 u'tim',
 u'howard',
 u'is',
 u'cursing',
 u'away',
 u'at',
 u'us',
 u'ahead',
 u'of',
 u'sunday',
 u'#LFC']

## Clean GloVe pre-train Twitter word vector
the original word vector is 257.7MB for only 25 dimension  
it contains a lots of non english words that we don't need in this task  
here we extract only the enlish words and < user >, < link >, etc to speed up the task

In [22]:
ENG_RE = re.compile(r"^[a-zA-Z]+$")
print ENG_RE.match(r"i")
print ENG_RE.match(r"you")
print ENG_RE.match(r"somethejrwekewr")
print ENG_RE.match(r"ayyaşlar")
print ENG_RE.match(r"ﾖｲｼｮｯ")
print ENG_RE.match(r"한강대교")
print ENG_RE.match(r"nاحترم")
line = "ciğerimiz -1.5423 -0.6015 1.4372 -0.81059 -1.4139 -0.238 -1.1359 -1.5304 -0.81756 0.023343 -1.0026 -0.95681 1.1536 1.2247 0.75876 -0.21646 0.98548 1.2462 0.61908 -0.45587 -0.054105 0.68484 -0.040623 0.04541 -1.0274"
print ENG_RE.match(line.split(" ", 1)[0])
line = "nnこれをこうして -1.8114 0.31154 -1.804 2.3232 -1.4398 -0.34702 -0.69259 -0.90919 1.7543 -0.47194 -1.6467 -3.5453 2.3154 1.0312 -3.6022 -2.2001 -4.8098 -2.9928 0.99323 -1.2025 -0.37103 -4.3479 1.6415 -1.3526 0.3258"
print ENG_RE.match(line.split(" ", 1)[0])

None
None
None
None
None
None


This is the code for creating the cleaned file. Can only be runned if glove.twitter.27B.25d.txt is downloaded.  
for twitter from glove.twitter.27B.zip http://nlp.stanford.edu/projects/glove/

Data downloaded form http://nlp.stanford.edu/data/glove.twitter.27B.zip  
        (2B tweets, 27B tokens, 1.2M vocab, uncased, 25d, 50d, 100d, & 200d vectors, 1.42 GB download)
        The extracted file must be placed in the Lexicon_And_WE sub folder at the same folder of this file
        And make sure the file path in the code exist

In [23]:
# commented out because can only be run if glove.twitter.27B/glove.twitter.27B.25d.txt exist 

# import re

# ENG_RE = re.compile(r"^[a-zA-Z]+$")
# GLOVE_25d = 'Lexicon_And_WE/glove.twitter.27B/glove.twitter.27B.25d.txt'

# target = open('Lexicon_And_WE/glove.twitter.27B/glove.twitter.25d-cleaned-en.txt', 'w')
# with open(GLOVE_25d) as gfile:
#     for line in gfile:
#         word = line.split(' ', 1)[0] # split first occurance
#         if ENG_RE.match(word):
#             target.write(line)
# target.close()

In [24]:
# test WE_GloVe_Twitter
import Preprocessor
import PipelineRunner

reload(Preprocessor)
gold_set = PipelineRunner.getTrainingSetB("Exercise2_data/twitter-dev-gold-B.tsv", Preprocessor.tokenise)
lexi = Preprocessor.WE_GloVe_Twitter().transform(gold_set['tokens'])

In [25]:
print len(lexi)
lexi[:2]

1374


[[-0.12192394117647061,
  0.23052739999999999,
  -0.30798617647058829,
  -0.33501582352941173,
  -0.068645552941176469,
  0.0093505294117647107,
  1.0058978823529412,
  -0.096249941176470594,
  -0.29775958823529408,
  -0.24241999999999997,
  -0.05634823529411765,
  0.16948582352941177,
  -4.0174423529411767,
  0.41108117647058823,
  0.39450341764705887,
  0.080877647058823526,
  0.034647176470588251,
  -0.29263617647058821,
  -0.4714964117647058,
  -0.0014791176470588057,
  -0.41170829999999997,
  0.020581000000000002,
  -0.024761117647058819,
  -0.23383629411764706,
  -0.089179529411764699,
  -1.2971,
  -0.70268,
  -2.3364,
  -1.5596,
  -0.85091,
  -1.2542,
  -0.098218,
  -1.2279,
  -1.2761,
  -1.3655,
  -1.6672,
  -2.501,
  -6.3371,
  -0.41398,
  -0.5531,
  -0.62011,
  -0.71249,
  -1.2579,
  -2.5867,
  -1.3695,
  -1.4068,
  -1.1655,
  -0.94472,
  -1.6099,
  -0.90682,
  1.5649,
  0.85265,
  0.7974,
  0.87498,
  1.4854,
  1.3908,
  2.2782,
  2.3866,
  0.42165,
  0.74383,
  0.77335,
  1